Scripts/DCS-BIOS/lib/Protocol.lua

```diff
@@ -76,75 +77,64 @@ function BIOS.protocol.endModule()
 			for identifier, args in pairs(category) do
 				local outputs = args.outputs or {}
 				for _, output in pairs(outputs) do
-					local full_identifier = BIOS.util.addressDefineIdentifier(moduleName, identifier)
+					local address_identifier = output.address_identifier or BIOS.util.addressDefineIdentifier(moduleName, identifier)
 					local addressStr = output.address and string.format("0x%X", output.address) or ""
 					local maskStr = output.mask and string.format("0x%X", output.mask) or ""
 					local shiftByStr = output.shift_by and tostring(output.shift_by) or ""
 
 					-- Define line with address, mask, and shiftby
-					local line = "#define " .. full_identifier .. " " .. addressStr
+					local line = "#define " .. address_identifier .. " " .. addressStr
 					if maskStr ~= "" then line = line .. ", " .. maskStr end
 					if shiftByStr ~= "" then line = line .. ", " .. shiftByStr end
-
-					if not existingDefines[full_identifier] then
-						table.insert(lineOrder, full_identifier)
+	
+					if not addresses[address_identifier] then
+						addresses[address_identifier] = line
 					end
-
-					existingDefines[full_identifier] = line
-
+	
 					-- Additional line with only the address and _ADDR suffix
 					if addressStr ~= "" then
-						local addressOnlyIdentifier = full_identifier .. "_ADDR"
+						local addressOnlyIdentifier = address_identifier .. "_ADDR"
 						local addressOnlyLine = "#define " .. addressOnlyIdentifier .. " " .. addressStr
 
-						if not existingDefines[addressOnlyIdentifier] then
-							table.insert(lineOrder, addressOnlyIdentifier)
+						if not addresses[addressOnlyIdentifier] then
+							addresses[addressOnlyIdentifier] = addressOnlyLine
 						end
-
-						existingDefines[addressOnlyIdentifier] = addressOnlyLine
 					end
 				end
 			end
 		end
-
-		-- Write the updated content to the file
-		local address_header_file, err = io.open(path, "w")
-		if err then
-			print("Error opening file:", err) -- Print error if unable to open file
-			return
-		end
-		for _, identifier in ipairs(lineOrder) do
-			address_header_file:write(existingDefines[identifier] .. "\n")
-		end
-		address_header_file:close() -- Close the header file
 	end
 	pcall(saveDoc)
 	pcall(saveAddresses)
 	moduleBeingDefined = nil
 	end
 end
+function BIOS.protocol.writeAddressesFile()
+	-- Sort the identifiers before writing
+	local sortedIdentifiers = {}
+	for identifier in pairs(addresses) do
+		table.insert(sortedIdentifiers, identifier)
+	end
+	table.sort(sortedIdentifiers)
+
+	-- Write the updated content to the file
+	local address_header_file, err = io.open(lfs.writedir()..[[Scripts\DCS-BIOS\doc\Addresses.h]], "w")
+	if err then
+		print("Error opening file:", err) -- Print error if unable to open file
+		return
+	end
+	for _, identifier in ipairs(sortedIdentifiers) do
+		address_header_file:write(addresses[identifier] .. "\n")
+	end
+	address_header_file:close() -- Close the header file
+
+	addresses = nil -- Release the memory

```

Comment:
Where is this defined? Is this a global that I'm not seeing somewhere? We're trying to move towards a place where everything is locally-scoped. Ideally this function wouldn't need to define or reference anything new outside of it.

If I'm following right, the current behavior is to add to a table of addresses every time a module is exported, and then write that table to a file on `init()` - is that right? If so, can `writeAddressesFile()` just iterate over `exportModules` to get all the data it needs there instead? That way everything relating to the arduino file is kept in one place, and we're reusing an existing variable that already has the data. It should make refactoring easier down the line as well.

Reply:
Yes, thats a global to avoid re-reading and -writing the Addresses.h for every module. I'm not familiar enough with DCS-BIOS to know how to iterate over the modules properly and get everything. If you can point me in the right direction, i will refactor it to make it self-contained.

Reply:
It's not a huge deal.

At the top of the file you'll see a few variables like so:
```lua
local exportModules = {}
local aircraftNameToModuleNames = {}
local aircraftNameToModules = {}
local lastAcftName = ""
```

if you could just add `local addresses = {}` to the end of that block so at least the variable is scoped to that file, it will help.

Reply:
Thats where it is :)

https://github.com/DCS-Skunkworks/dcs-bios/pull/240/files/d43da00f59870d7d4e185efca253ab39866388e7#diff-6cefe804a43126b13d284b7c366c4675a3bb50f5d49a54c4de7ea5d577b89a12R9